### PR TITLE
[Feature] UniswapV2PairConfig.sol

### PR DESCRIPTION
**Pull request by swc0620**
# Overview
UniswapV2PairConfig.sol implements all the premises required to run and test the features of kick-finance. In reality, I can simply assume that every backer to kick-finance holds sufficient amount of stable coins(e.g. DAI) and there already exists a liquidity pool in UniswapV2 regarding that stable coin. However, in the toy-project condition like this, these premises do not hold. This means that before I implement new features such as backProject(), I must first set the following basic conditions on which my toy-project can work.

1. creates a fake stable coin(BT, backingToken), which is used by backers when they back the project, and a pair token(AT, auxiliaryToken) which is used when creating UniswapV2 liquidity pool
2. distributes newly created fake stable coin to proposer and backers
3. creates BT-AT liquidity pool on UniswapV2
4. lets the proposer add liquidity to BT-AT liquidity pool in advance, so when new backings are made, swaps can be done on the BT-AT liquidity pool

UniswapV2PairConfig.sol solely focuses on setting these conditions. ProjectFactory.sol is a factory contract which ought to be provoked by the proposer to create new Project he or she would like to propose.

### 1) MockDAIToken.sol
```
function faucet() external
```
- [X] inherits ERC-20
- [X] creates faucet() function which is used in initial distribution of MDAI
### 2) MockDAIToken.test.ts
- [X] checks whether the contract has given data on construction
- [X] checks whether the MDAI can be successfully distributed via faucet() function

### 1) MockWETHToken.sol
```
function faucet() external
```
- [X] inherits ERC-20
- [X] creates faucet() function which is used in initial distribution of MWETH
### 2) MockDAIToken.test.ts
- [X] checks whether the contract has given data on construction
- [X] checks whether the MWETH can be successfully distributed via faucet() function

## 3. ProjectFactory
```
function createProject(string memory description_, uint minimumBacking_, address manufacturer_) external
function getProjects() public view returns (address[] memory)
```
### 1) ProjectFactory.sol
- [X] creates new project contract and store its address
- [x] returns an array of addresses of created projects
### 2) ProjectFactory.test.ts
- [X] checks whether new project contract is successfully created and its address is stored
- [X] checks whether addresses of created projects are returned

## 4. UniswapV2Config
```
function addLiquidity(address backingToken_, address auxiliaryToken_, address router_, uint amountBT_, uint amountAT_) external
```
### 1) UniswapV2Config.sol
- [X] transfers MDAI, MWETH from the proposer to UniswapV2Config.sol
- [X] approves UniswapV2Router02 to manage transferred MDAI, MWETH
- [X] creates liquidity pool of BT-AT pair via UniswapV2Router02
- [X] transfers MDAI, MWETH from UniswapV2Config.sol to BT-AT pair via UniswapV2Router02, thus provide liquidity to BT-AT liquidity pool
### 2) UniswapV2Config.test.ts
- [X] checks whether the contract has given data on construction
- [X] lets the proposer receive initial MDAI, MWETH via faucet() function
- [X] approves UniswapV2Config.sol to manage received MDAI, MWETH and provoke addLiquidity() function
- [X] checks whether anticipated Transfer events are emitted
- [X] checks whether the balance of LP token of UniswapV2Config.sol is increased on BT-AT liquidity pool

### CODE REVIEW

**Review by jaeboklee-modo — COMMENTED** [review]
I'm glad to review your first solidity project.

I'm curious about what is the purpose of using a mainnet forking. The only forked block state that this test code uses is the uniswap router, which doesn't need to be since it is independent of the block state. 

If you intended to use the uniswap pair's state, then forking would be a good option, but the router contract itself doesn't need to be the same as the state of the mainnet because all the pool in the uniswap is independent. 

Unnecessary forking can make the test slower and restrict the test environment than deploying itself. 

Therefore could you clarify the intention of the mainnet forking in your test environment?

File: contracts/Project.sol [hunk 1]
```diff
@@ -13,25 +22,65 @@ contract Project {
...
+        // let this contract be in control of amountBT
+        MockDAIToken mockDAIToken = MockDAIToken(backingToken);
+        MockWETHToken mockWETHToken = MockWETHToken(auxiliaryToken);
+        uint amountBT = amountBT_ * 10 ** mockDAIToken.decimals();
+        require(mockDAIToken.transferFrom(msg.sender, address(this), amountBT), "transferFrom failed.");
+    
+        // approve UniswapV2Rounter to withdraw amountBT
+        require(mockDAIToken.approve(0x7a250d5630B4cF539739dF2C5dAcb4c659F2488D, amountBT), "approve failed.");
+        
+        // swap BT-AT
```
[the remark this answers is not in the document]
Comment (swc0620) [reply contracts/Project.sol:69]: 👍

File: test/ProjectFactory.test.ts [hunk 2]
```diff
@@ -22,12 +22,12 @@ describe("ProjectFactory", () => {
 
     context("when new ProjectFactory is deployed", async() => {
```
[the remark this answers is not in the document]
Comment (swc0620) [reply test/ProjectFactory.test.ts:23]: Yes I figured this out later, and pushed pr#3 to solve this

File: contracts/Project.sol [hunk 3]
```diff
@@ -13,25 +22,65 @@ contract Project {
...
         backingTime.open = true;
         backingTime.closeTime = block.timestamp + backingDuration_;
+
+        pairAddress = pairAddress_;
+        backingToken = backingToken_;
+        auxiliaryToken = auxiliaryToken_;
+    }
+
+    function backProject(uint amountBT_) external payable {
+        require(backingTime.open == true);
```
[the remark this answers is not in the document]
Comment (swc0620) [reply contracts/Project.sol:56]: In order to restrict this pr to only manage UniswapV2PairConfig.sol and ProjectFactory.sol, I added error comments to these require clauses on new branch(feature/Project/backProject).

File: contracts/Project.sol [hunk 4]
```diff
@@ -13,25 +22,65 @@ contract Project {
...
+        require(msg.value > minimumBacking);
+
+        // let this contract be in control of amountBT
+        MockDAIToken mockDAIToken = MockDAIToken(backingToken);
+        MockWETHToken mockWETHToken = MockWETHToken(auxiliaryToken);
+        uint amountBT = amountBT_ * 10 ** mockDAIToken.decimals();
+        require(mockDAIToken.transferFrom(msg.sender, address(this), amountBT), "transferFrom failed.");
+    
+        // approve UniswapV2Rounter to withdraw amountBT
+        require(mockDAIToken.approve(0x7a250d5630B4cF539739dF2C5dAcb4c659F2488D, amountBT), "approve failed.");
```
[the remark this answers is not in the document]
Comment (swc0620) [reply contracts/Project.sol:67]: Yes. I resolved this in the new branch(feature/Project/backProject).

File: test/UniswapV2PairConfig.test.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
...utability":"nonpayable","type":"function"},{"constant":true,"inputs":[],"name":"token0","outputs":[{"internalType":"address","name":"","type":"address"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[],"name":"token1","outputs":[{"internalType":"address","name":"","type":"address"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[],"name":"totalSupply","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[{"internalType":"address","name":"to","type":"address"},{"internalType":"uint256","name":"value","type":"uint256"}],"name":"transfer","outputs":[{"internalType":"bool","name":"","type":"bool"}],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":false,"inputs":[{"internalType":"address","name":"from","type":"address"},{"internalType":"address","name":"to","type":"address"},{"internalType":"uint256","name":"value","type":"uint256"}],"name":"transferFrom","outputs":[{"internalType":"bool","name":"","type":"bool"}],"payable":false,"stateMutability":"nonpayable","type":"function"}];
```
[the remark this answers is not in the document]
Comment (swc0620) [reply test/UniswapV2PairConfig.test.ts:76]: I made a new directory containing only abis

**Comment by swc0620** [comment]
> I'm glad to review your first solidity project.
> 
> I'm curious about what is the purpose of using a mainnet forking. The only forked block state that this test code uses is the uniswap router, which doesn't need to be since it is independent of the block state.
> 
> If you intended to use the uniswap pair's state, then forking would be a good option, but the router contract itself doesn't need to be the same as the state of the mainnet because all the pool in the uniswap is independent.
> 
> Unnecessary forking can make the test slower and restrict the test environment than deploying itself.
> 
> Therefore could you clarify the intention of the mainnet forking in your test environment?

yes, sounds very persuasive. I am not sure and am not aware of how slow running test codes on forked mainnet is compared to running them on newly deployed contracts on local network is, but the reasons I used mainet forking are as follow.

1. There was a lack of gas issue. First I tried to use default wallets provided by waffle(waffle.provider.getWallets()), but lack of gas error popped out, so I used ethereum holding wallets(ethers.getSigner()) from mainnet instead and this solved the issue.
2. I used more Uniswap contracts on pr#4, including UniswapV2Factory and UniswapV2Router02. I was more convenient to used them via mainnet forking, but none of them typically demands use of mainnet "state", so this was just a pinch of convenience added.
3. This is a toy-project thus it seemed quite challenging to write test code on mainnet forked environment.

**Comment by swc0620** [comment]
pr#4 is branched off from pr#1 and I applied all the code modifications mentioned here on pr#4, so i will close this pr and continue there

**Pull request closed without merging**